Pre-processing of Card UI to reduce the future work conflicts between partial auth and UX improvements on onboarding page

eason9487 opened the pull request:
### Changes proposed in this Pull Request:

Part of #1000

- Add `disabled` and `alignIcon` props to `<AccountCard>` for styling
- Replace the implementations within `<GoogleAccountCard>` by `<AccountCard>`
- Merge the Cards of WordPress (JetPack) account and Google account into the same Section
- Move the applying of disabled style down to the Card layer in the `<GoogleMCAccount>`
- Remove the dependency of `<GoogleAccount>` within **js/src/setup-ads/ads-stepper/setup-accounts/index.js**

The requirements of UX improvements can be found at pcTzPl-6u

### Additional notes

- Related discussion #1030 hasn't resulted in a conclusion. But eventually, we need more custom styles in the `<AccountCard>`, so I put the class names onto some internal elements in advance by the existed naming convention in the codebase. Such as the usages in the [app-tab-nav](https://github.com/woocommerce/google-listings-and-ads/blob/1.5.1/js/src/components/app-tab-nav/index.js#L32-L46)
- The main goal of this PR is to reduce the chance of conflict between the two projects. There will be some temporary inconsistency with other components on the page, which will be adjusted later in the subsequent development.
- Some inconsistencies between two projects' Figma and the current codebase were confirmed with Jill:
    1. The permission description of Google account card should use the **UX improvements** version
    2. The title of Google account card should be "Google account", which is the same title on the re-connection page

### Screenshots:

💡  The adjusted UI of WordPress and Google account cards is upon the red line, and the one under red line is the original UI.

#### 📷 . Disabled card/secion
![2021-10-18 16 34 19](https://user-images.githubusercontent.com/17420811/137718534-9c7eeef3-37bb-4b65-968f-f48adcf33aaa.png)

#### 📷 . Not yet connected
![2021-10-18 16 35 29](https://user-images.githubusercontent.com/17420811/137718539-7dc8dd25-f81e-462f-9c95-c108091e5537.png)

#### 📷 . Connected
![2021-10-18 16 36 41](https://user-images.githubusercontent.com/17420811/137718551-13ac3914-edd6-4f45-99e9-49c57be7b7f2.png)

### Detailed test instructions:

1. Go to the onboarding page.
2. The function of connecting Google account should work as before.
3. Although it may not be easy to evaluate because I can't fully describe my later implementation, and I don't know the implementation design that Gan will take. Nevertheless, it should be expected that it would be easier to continue development of both projects and less conflict chance with this adjustment. 🤞 

### Changelog entry

eason9487 committed:
Add disabled and alignIcon props to AccountCard for styling

eason9487 committed:
Replace the GoogleAccount component with AccountCard

eason9487 committed:
Rename google-account folder to google-account-card

eason9487 committed:
Merge the WordPressDotComAccount and GoogleAccountCard into a Section

eason9487 committed:
Replace the implementation of GoogleAccountCard with AccountCard

eason9487 committed:
Don't wrap the Section within DisabledDiv for GoogleMCAccount

eason9487 commented on js/src/components/google-account/card-content.js:
```diff
@@ -1,44 +0,0 @@
-		return <ConnectAccount disabled={ disabled } />;
-	}
-
-	if ( ! google ) {
-		return (
-			<TitleButtonLayout
-				title={ __(
-					'Error while loading Google account info',
-					'google-listings-and-ads'
-				) }
-			/>
-		);
-	}
```
I didn't move this part to the new implementation, because with the previous continued development, it has been no longer entering this logic.

eason9487 commented on js/src/wcdl/subsection/title/index.scss:
```diff
@@ -6,5 +6,8 @@
+
+	&:not(:empty) {
+		margin-bottom: 8px;
+	}
```
To remove the extra 8px height when rendering the loading spinner only at 

https://github.com/woocommerce/google-listings-and-ads/blob/92b08874adc1084b2ccde38954e10787355251d5/js/src/components/google-account-card/google-account-card.js#L13-L15

tomalec replied:
:nail_care: :scroll: I'm not sure if we discussed it already, but if we'd make Cards body a `display: flex; flex-direction: column; gap: 8px;` we would not have to bother about `:not(:empty)` and `:last-child` manual margin juggling, but would get that collapsing for empty and edge items for free.

If we are about to stay with manual reset, I'd add a code comment that it's to collapse the margin for empty titles. So we won't forget this in the future.

eason9487 replied:
A similar discussion is here https://github.com/woocommerce/google-listings-and-ads/pull/1019#discussion_r714246675.

In the `<Subsection.Title>` case, currently its parent elements are:
- `<Section.Card.Body>`
- `<FlexBlock>`
- `<Subsection>`
- `<ContentButtonLayout>`
- `<div>`

If transfer the space style to upper layers, the changing scope will be much larger. I lean toward not doing so for now. Added comment by https://github.com/woocommerce/google-listings-and-ads/pull/1052/commits/a41336cb2e3a65e88724a9b9f292677500b65001

eason9487 commented on js/src/components/google-account-card/auth-google-account-card.js:
```diff
@@ -0,0 +1,87 @@
+		<AppDocumentationLink
+			context="setup-mc-accounts"
+			linkId="required-google-permissions"
+			href="https://docs.woocommerce.com/document/google-listings-and-ads/#required-google-permissions"
```
The anchor should work after finishing #1053

eason9487 committed:
Update e2e test for the step-1-accounts of setup-mc

tomalec commented on js/src/components/google-account-card/auth-google-account-card.js:
```diff
@@ -0,0 +1,87 @@
+
+	const handleConnectClick = async () => {
+		try {
+			const d = await fetchGoogleConnect();
```
:nail_care:  Would you mind changing the `d` to a more descriptive name?

tomalec replied:
ah, it was like this before, sorry I missed that, as GH didn't show the file move. We could leave it as is, to emphasize just the move, or improve by the way.

eason9487 replied:
I tried to keep the record for this file by `git mv`, but it still result in delete and addition records. 😞  Maybe there is too much diff in the later commit. Changed by https://github.com/woocommerce/google-listings-and-ads/pull/1052/commits/8523b5e928360b234af3ec1f4b9d998814891df3.

tomalec commented on js/src/components/google-account-card/auth-google-account-card.js:
```diff
@@ -0,0 +1,87 @@
+							'You will be prompted to give WooCommerce access to your Google account. Please check all the checkboxes to give WooCommerce all required permissions. <link>Read more</link>',
+							'google-listings-and-ads'
+						),
+						conversionMap
```
:nail_care::nail_care: Wouldn't it be cleaner to put the link here directly, so we would not need `conversionMap` at all and a more declarative block in a single place (without the need to jump around to find the content of this var).
```jsx
link: (
		<AppDocumentationLink
```

When I get here, I silly thought sparked in my head: "conversion? what kind of conversion we do when connecting accounts?" :see_no_evil:  (probably I got stuck for too long with currencies problems ;))

eason9487 replied:
I took the variable name from [its document](https://github.com/WordPress/gutenberg/blob/%40wordpress/components%4012.0.8/packages/element/src/create-interpolate-element.js#L105). 🙈  The "Read more" link will be reused in another description later, so I placed it here in advance. I adjusted the parameter of `createInterpolateElement` by https://github.com/woocommerce/google-listings-and-ads/pull/1052/commits/1d78fa610258fccae7222b3ae59c9082b26eb124 and hope this could make it clearer.

![image](https://user-images.githubusercontent.com/17420811/138216300-11389878-0879-4958-a65c-98c4d10f1662.png)

tomalec commented on js/src/components/google-account-card/auth-google-account-card.js:
```diff
@@ -0,0 +1,87 @@
+			appearance={ APPEARANCE.GOOGLE }
+			disabled={ disabled }
+			hideIcon
+			alignIcon="top"
```
:grey_question: :scroll: 
Do we need `alignIcon` when it's `hideIcon`.
Also maybe we can somehow marge those two props together, to avoid such confusion.
Like, `iconPosition = top | center | hidden`?

eason9487 replied:
Sorry for not stating it clearly first. 🙏  By the **UX Improvements** project, all account cards will have a leading icon. But I only adjusted Google account card for the **Partial oauth** project, and other account cards stay with the original visual results. To make it consistent as possible, I was thinking of hiding the icon here, and Gan can just remove all `hideIcon` when he is working on UX Improvements.

// cc @ecgan

tomalec commented on js/src/components/google-account-card/index.js:
```diff
@@ -0,0 +1 @@
+export { default } from './google-account-card';
```
:sunglasses: nice to have `index.js` beeing just an index and have the content in named files.

eason9487 replied:
I like this way of file structure. 👍

tomalec commented on js/src/components/google-account-card/google-account-card.js:
```diff
@@ -0,0 +1,22 @@
+	const { google, hasFinishedResolution } = useGoogleAccount();
+
+	if ( ! hasFinishedResolution ) {
+		return <AccountCard appearance={ {} } description={ <AppSpinner /> } />;
```
:scroll: It wasn't here before, but we could consider making this element A11y friendly with indicating loading state

tomalec replied:
:nail_care:  :scroll:  Also, maybe we can consider introducing a default behavior for no appearance, so we would not have to provide an empty object here.

eason9487 replied:
Since the **UX Improvements** project will have more interactive adjustments to all account cards on the onboarding page, and I don't know how Gan will do the design. So I don't aim to make detailed changes in this PR, but to stick to the original implementation as much as possible. And we could leave this 📜  as a dev note for Gan. 😎 

// cc @ecgan

tomalec commented on js/src/components/account-card/index.scss:
```diff
@@ -1,4 +1,20 @@
+
+	&__description {
+		p:last-child {
+			margin-bottom: 0;
```
:construction: I think now, we can remove the same rule StoreAddressCard from https://github.com/woocommerce/google-listings-and-ads/blob/tweak/1000-preprocess-card-ui/js/src/components/contact-information/store-address-card.scss#L20:L20

:nail_care: Also here, I'd consider making `&__description{ display: flex; flex-direction: column; gap: x }` Is this space is rather a property of the container than anyhow specific to the paragraph.

eason9487 replied:
Changed along with other styles that can be replaced. https://github.com/woocommerce/google-listings-and-ads/pull/1052/commits/64e371d6f4fa5538fc715e021db79923f32a569b

tomalec commented on js/src/components/google-account-card/google-account-card.js:
```diff
@@ -0,0 +1,22 @@
+		return <ConnectedGoogleAccountCard googleAccount={ google } />;
+	}
+
+	return <AuthGoogleAccountCard disabled={ disabled } />;
```
:grey_question: :nail_car: Is the `Auth` stands here for `Authorised` or `Authorization`?
If it's "Authorization", to me `GoogleAuthorizationCard` sounds more natural that `AuthorizationGoogleCard`, but I'm not born to English or expert.

eason9487 replied:
It stands for "Authorize", and the former name of this component was `ConnectAccount`. After adding the partial oauth feature, the account state could be:

1. unconnected
2. connected with partial permissions
3. connected with all needed permissions.

I plan to handle all of them in this component. So, in the latter two states, the next action is no longer "connect", but continues to ask for permissions, so I renamed it to the action "Auth" that summarizes all states. 

Renamed to `AuthorizeGoogleAccountCard ` by https://github.com/woocommerce/google-listings-and-ads/pull/1052/commits/832433c656433f4df46a74e3b422d9dc3bf4e1b9.

tomalec commented on js/src/setup-ads/ads-stepper/setup-accounts/index.scss:
```diff
@@ -0,0 +1,7 @@
+.gla-setup-ads-accounts {
+	&__google-card {
+		.gla-account-card__description {
+			color: $gray-700;
```
:grey_question: We have a similar rule, for the contact information notice https://github.com/woocommerce/google-listings-and-ads/blob/tweak/1000-preprocess-card-ui/js/src/components/contact-information/contact-information-preview-card.scss#L12:L12

I wonder if we could somewhat merge those two. And got some semantic description of the state when the description should be black or grey.

eason9487 replied:
I'm not sure if there would be more and more customization styles of font typography. So I tend to wait until the **UX Improvements** project is (almost) complete and then re-examining how we can extract their patterns and internalize them into the `<AccountCard>` component.

tomalec reviewed (changes requested):
Reviewed the code, tested locally, the MC setup, Ads setup, and settings cards. LGTM besides one note about no-longer-needed code left, plus a number of cosmetic ideas.

tomalec reviewed (approved):
pre-approving, as I guess there are still PRs planned that may address all remaining issues.

eason9487 committed:
Change a variable to a more descriptive name.

Address https://github.com/woocommerce/google-listings-and-ads/pull/1052#discussion_r732737425

eason9487 committed:
Adjust the parameter of createInterpolateElement to make it clearer.

Address https://github.com/woocommerce/google-listings-and-ads/pull/1052#discussion_r732741140

eason9487 committed:
Replace some StoreAddressCard styles with AccountCard's styles and props.

Address https://github.com/woocommerce/google-listings-and-ads/pull/1052#discussion_r733025244

eason9487 committed:
Add a comment for a css rule.

For https://github.com/woocommerce/google-listings-and-ads/pull/1052#discussion_r732906822

eason9487 committed:
Rename AuthGoogleAccountCard to AuthorizeGoogleAccountCard.

Address https://github.com/woocommerce/google-listings-and-ads/pull/1052#discussion_r733029125
